pkg/services/ngalert/schedule/schedule.go

```diff
@@ -230,12 +220,16 @@ func (sch *schedule) updateRulesMetrics(alertRules []*ngmodels.AlertRule) {
 	sch.metrics.SchedulableAlertRulesHash.Set(float64(hashUIDs(alertRules)))
 }
 
-func (sch *schedule) processTick(ctx context.Context, dispatcherGroup *errgroup.Group, tick time.Time) ([]readyToRunItem, map[ngmodels.AlertRuleKey]struct{}) {
+func (sch *schedule) processTick(ctx context.Context, dispatcherGroup *errgroup.Group, tick time.Time) ([]readyToRunItem, map[ngmodels.AlertRuleKey]struct{}, []readyToUpdateItem) {
 	tickNum := tick.Unix() / int64(sch.baseInterval.Seconds())
 
-	if err := sch.updateSchedulableAlertRules(ctx); err != nil {
+	// update the local registry. If there was a difference between the previous state and the current new state, rulesDiff will contains keys of rules that were updated.
+	rulesDiff, err := sch.updateSchedulableAlertRules(ctx)
```

Comment:
Overall, all this looks generally good - my only concern (which is something we can worry about in the future) is:

The fact that the rules diff contains only previously existing rules that were updated feels like a loaded foot-gun to me. It's documented clearly on the `diff` struct, but not in other places.

I understand that _within this PR_ it's not a problem, as the only thing reading this data needs to perform different actions on new vs. updated anyway, and handles new through a completely different path. Still, I think it's an easy opportunity for future bugs where new rules get ignored because another developer thinks the diff contains everything that changed.

At a minimum, I wonder if we could propagate that comment that says it's only updated rules in more spots, like the doc of `updateSchedulableAlertRules` and the registry's `.set()`? Or, we could refer to this object as `reloadedRules` to imply that it's existing rules that changed, not new ones?

Again, super nitpicky - and we could address it in a future non-backported PR if we prefer.

Reply:
Left this comment before the final commit. The `updated := rulesDiff.updated` helps a lot :+1: